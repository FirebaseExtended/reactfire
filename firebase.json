{
  "emulators": {
    "auth": {
      "port": 9099
    },
    "functions": {
      "port": 5001
    },
    "firestore": {
      "port": 8080
    },
    "database": {
      "port": 9000
    },
    "storage": {
      "port": 9199
    },
    "ui": {
      "enabled": true
    }
  },
  "storage": {
    "rules": "storage.rules"
  }
}
