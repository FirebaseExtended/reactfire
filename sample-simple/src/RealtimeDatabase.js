import React, { useState } from 'react';
import {
  AuthCheck,
  SuspenseWithPerf,
  useDatabaseList,
  useDatabaseObject,
  useDatabase
} from 'reactfire';

const Counter = props => {
  const database = useDatabase();
  const ref = database().ref('counter');
  const increment = amountToIncrement => {
    ref.transaction(counterVal => {
      return counterVal + amountToIncrement;
    });
  };

  const { snapshot } = useDatabaseObject(ref);
  const counterValue = snapshot.val();

  return (
    <>
      <button onClick={() => increment(-1)}>-</button>
      <span> {counterValue} </span>
      <button onClick={() => increment(1)}>+</button>
    </>
  );
};

const AnimalEntry = ({ saveAnimal }) => {
  const [text, setText] = useState('');
  const [disabled, setDisabled] = useState(false);

  const onSave = () => {
    setDisabled(true);
    saveAnimal(text).then(() => {
      setText('');
      setDisabled(false);
    });
  };

  return (
    <>
      <input
        value={text}
        disabled={disabled}
        placeholder="Iguana"
        onChange={({ target }) => setText(target.value)}
      />
      <button onClick={onSave} disabled={disabled || text.length < 3}>
        Add new animal
      </button>
    </>
  );
};

const List = props => {
  const database = useDatabase();
  const ref = database().ref('animals');
  const changes = useDatabaseList(ref);
  const addNewAnimal = commonName => {
    const newAnimalRef = ref.push();
    return newAnimalRef.set({
      commonName
    });
  };

  const removeAnimal = id => ref.child(id).remove();

  return (
    <>
      <AnimalEntry saveAnimal={addNewAnimal} />
      <ul>
        {changes.map(({ snapshot }) => (
          <li key={snapshot.key}>
            {snapshot.val().commonName}{' '}
            <button onClick={() => removeAnimal(snapshot.key)}>X</button>
          </li>
        ))}
      </ul>
    </>
  );
};

const SuspenseWrapper = props => {
  return (
    <SuspenseWithPerf fallback="loading..." traceId="RTDB-root">
      <AuthCheck fallback="sign in to use Realtime Database">
        <h3>Sample Object Listener</h3>
        <Counter />
        <h3>Sample List Listener</h3>
        <List />
      </AuthCheck>
    </SuspenseWithPerf>
  );
};

export default SuspenseWrapper;
