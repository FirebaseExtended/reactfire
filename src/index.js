import createAppContainer from './createAppContainer';
import createAuthContainer from './createAuthContainer';

export {
  createAppContainer,
  createAuthContainer,
};
