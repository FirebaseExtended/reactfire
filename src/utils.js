/**
 * Returns whether or not the provided input is a JavaScript function.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a JavaScript function.
 */
export function isFunction(input) {
  return typeof input === 'function';
}

/**
 * Returns whether or not the provided input is a non-null JavaScript object.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a non-null JavaScript object.
 */
export function isNonNullObject(input) {
  return Object.prototype.toString.call(input) === '[object Object]';
}

/**
 * Returns whether or not the provided input is a Firebase Database Reference.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a Firebase Database Reference.
 */
export function isDatabaseReference(input) {
  return isNonNullObject(input) &&
    typeof input.ref === 'object' &&
    typeof input.on === 'function' &&
    typeof input.set === 'function' &&
    typeof input.once === 'function' &&
    typeof input.transaction === 'function';
}

/**
 * Returns whether or not the provided input is a Firebase Database Query.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a Firebase Database Query.
 */
export function isDatabaseQuery(input) {
  return isNonNullObject(input) &&
    typeof input.ref === 'object' &&
    typeof input.on === 'function' &&
    typeof input.once === 'function' &&
    typeof input.endAt === 'function' &&
    typeof input.orderByChild === 'function';
}

/**
 * Returns whether or not the provided input is either a Firebase Database Reference or a Firebase
 * Database Query.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is either a Firebase Database Reference or
 * a Firebase Database Query.
 */
export function isDatabaseReferenceOrQuery(input) {
  return isDatabaseReference(input) || isDatabaseQuery(input);
}

/**
 * Returns whether or not the provided input is a Firebase App.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a Firebase App.
 */
export function isFirebaseApp(input) {
  return isNonNullObject(input) &&
    typeof input.name === 'string' &&
    typeof input.options === 'object' &&
    typeof input.delete === 'function';
}

/**
 * Returns whether or not the provided input is a React component.
 *
 * @param {any} input The input being tested.
 *
 * @return {boolean} Whether or not the provided input is a React component.
 */
export function isReactComponent(input) {
  return isNonNullObject(input) || isFunction(input);
}
