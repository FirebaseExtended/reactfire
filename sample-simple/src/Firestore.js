import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/performance';
import React, { useState } from 'react';
import {
  AuthCheck,
  SuspenseWithPerf,
  useFirestoreCollectionData,
  useFirestoreDocData,
  useFirebaseApp
} from 'reactfire';

const Counter = props => {
  const firebaseApp = useFirebaseApp();
  const ref = firebaseApp.firestore().doc('count/counter');

  const increment = amountToIncrement => {
    ref.update({
      value: firebaseApp.firestore.FieldValue.increment(amountToIncrement)
    });
  };

  const { value } = useFirestoreDocData(ref);

  return (
    <>
      <button onClick={() => increment(-1)}>-</button>
      <span> {value} </span>
      <button onClick={() => increment(1)}>+</button>
    </>
  );
};

const AnimalEntry = ({ saveAnimal }) => {
  const [text, setText] = useState('');
  const [disabled, setDisabled] = useState(false);

  const onSave = () => {
    setDisabled(true);
    saveAnimal(text).then(() => {
      setText('');
      setDisabled(false);
    });
  };

  return (
    <>
      <input
        value={text}
        disabled={disabled}
        placeholder="Iguana"
        onChange={({ target }) => setText(target.value)}
      />
      <button onClick={onSave} disabled={disabled || text.length < 3}>
        Add new animal
      </button>
    </>
  );
};

const List = props => {
  const firebaseApp = useFirebaseApp();
  const ref = firebaseApp.firestore().collection('animals');
  const animals = useFirestoreCollectionData(ref, { idField: 'id' });

  const addNewAnimal = commonName =>
    ref.add({
      commonName
    });

  const removeAnimal = id => ref.doc(id).delete();

  return (
    <>
      <AnimalEntry saveAnimal={addNewAnimal} />
      <ul>
        {animals.map(animal => (
          <li key={animal.id}>
            {animal.commonName}{' '}
            <button onClick={() => removeAnimal(animal.id)}>X</button>
          </li>
        ))}
      </ul>
    </>
  );
};

const SuspenseWrapper = props => {
  return (
    <SuspenseWithPerf fallback="loading..." traceId="firestore-demo-root">
      <AuthCheck fallback="sign in to use Firestore">
        <h3>Sample Doc Listener</h3>
        <SuspenseWithPerf
          fallback="connecting to Firestore..."
          traceId="firestore-demo-doc"
        >
          <Counter />
        </SuspenseWithPerf>
        <h3>Sample Collection Listener</h3>
        <SuspenseWithPerf
          fallback="connecting to Firestore..."
          traceId="firestore-demo-collection"
        >
          <List />
        </SuspenseWithPerf>
      </AuthCheck>
    </SuspenseWithPerf>
  );
};

export default SuspenseWrapper;
