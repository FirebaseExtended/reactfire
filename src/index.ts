import firebase from 'firebase/app';
import { SuspenseSubject } from './SuspenseSubject';

export type ReactFireGlobals = {
  _reactFireDatabaseCachedQueries: Array<firebase.database.Query>;
  _reactFireFirestoreQueryCache: Array<firebase.firestore.Query>;
  _reactFirePreloadedObservables: Map<string, SuspenseSubject<any>>;
};

export interface ReactFireOptions<T = unknown> {
  idField?: string;
  initialData?: T | any;
  suspense?: boolean;
}

export function checkOptions(options: ReactFireOptions, field: string) {
  // make sure the field passed in is a valid key of ReactFire Options
  if (field === 'idField' || field === 'initialData' || field === 'suspense') {
    return options ? (options[field] as ReactFireOptions['idField'] | ReactFireOptions['initialData'] | ReactFireOptions['suspense']) : undefined;
  }

  throw new Error(`Field "${field}" is not a valid key in ReactFireOptions`);
}

export function checkinitialData(options: ReactFireOptions) {
  return checkOptions(options, 'initialData');
}

export function checkIdField(options: ReactFireOptions) {
  return checkOptions(options, 'idField');
}

export * from './auth';
export * from './database';
export * from './firebaseApp';
export * from './firestore';
export * from './performance';
export * from './remote-config';
export * from './storage';
export * from './useObservable';
