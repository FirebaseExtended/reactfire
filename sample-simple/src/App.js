import React from 'react';
import AuthButton from './Auth';
import FirestoreCounter from './Firestore';
import Storage from './Storage';
import RealtimeDatabase from './RealtimeDatabase';
import {
  preloadFirestoreDoc,
  useFirebaseApp,
  preloadUser,
  preloadAuth,
  preloadFirestore,
  preloadDatabase,
  preloadStorage
} from 'reactfire';

const Fire = () => (
  <span role="img" aria-label="Fire">
    ðŸ”¥
  </span>
);

const Card = ({ title, children }) => {
  return (
    <div className="card">
      <h1>
        {title} <Fire />
      </h1>
      {children}
    </div>
  );
};

const preloadSDKs = firebaseApp => {
  return Promise.all([
    preloadFirestore(firebaseApp),
    preloadDatabase(firebaseApp),
    preloadStorage(firebaseApp),
    preloadAuth(firebaseApp)
  ]);
};

const preloadData = async firebaseApp => {
  const user = await preloadUser(firebaseApp);

  if (user) {
    preloadFirestoreDoc(
      firestore => firestore.doc('count/counter'),
      firebaseApp
    );
  }
};

const App = () => {
  const firebaseApp = useFirebaseApp();

  // kick off stuff we know our components will eventually need
  preloadSDKs(firebaseApp).then(preloadData(firebaseApp));

  return (
    <>
      <h1>
        <Fire /> ReactFire Demo <Fire />
      </h1>
      <div className="container">
        <Card title="Authentication">
          <AuthButton />
        </Card>

        <Card title="Firestore">
          <FirestoreCounter />
        </Card>

        <Card title="Storage">
          <Storage />
        </Card>

        <Card title="Realtime Database">
          <RealtimeDatabase />
        </Card>
      </div>
    </>
  );
};

export default App;
