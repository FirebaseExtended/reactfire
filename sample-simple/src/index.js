import React, { StrictMode } from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import * as serviceWorker from './serviceWorker';
import './app.css';
import { FirebaseAppProvider } from 'reactfire';
import 'firebase/performance';

const config = {
  apiKey: 'AIzaSyBg3u1sJlyJwQCE95oSDH_mtLABS-is8ZM',
  authDomain: 'rxfire-525a3.firebaseapp.com',
  databaseURL: 'https://rxfire-525a3.firebaseio.com',
  projectId: 'rxfire-525a3',
  storageBucket: 'rxfire-525a3.appspot.com',
  messagingSenderId: '844180061847',
  appId: '1:844180061847:web:400f7142e2d1aaeb'
};

ReactDOM.render(
  <StrictMode>
    <FirebaseAppProvider firebaseConfig={config} initPerformance>
      <App />
    </FirebaseAppProvider>
  </StrictMode>,
  document.getElementById('root')
); // https://reactjs.org/docs/strict-mode.html

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
