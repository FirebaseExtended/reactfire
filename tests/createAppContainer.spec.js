import React from 'react';
import renderer from 'react-test-renderer';
import * as firebase from 'firebase';

import { createAppContainer } from '../src/';

import { DummyComponent, mockFirebaseAppOptions } from './utils';


describe('createAppContainer()', () => {
  let defaultApp;

  beforeAll(() => {
    defaultApp = firebase.initializeApp(mockFirebaseAppOptions);
  });

  afterAll(() => firebase.app().delete());


  test('should throw given nothing as the first argument', () => {
    expect(() => {
      createAppContainer();
    }).toThrowErrorMatchingSnapshot();
  });

  test('should throw given an invalid React component as the first argument', () => {
    expect(() => {
      createAppContainer(null, defaultApp);
    }).toThrowErrorMatchingSnapshot();
  });

  test('should throw given nothing as the second argument', () => {
    expect(() => {
      createAppContainer(DummyComponent);
    }).toThrowErrorMatchingSnapshot();
  });

  test('should throw given an invalid Firebase App as the second argument', () => {
    expect(() => {
      createAppContainer(DummyComponent, null);
    }).toThrowErrorMatchingSnapshot();
  });

  test('should set "firebaseApp" prop to the provided Firebase App instance', () => {
    const WrappedDummyComponent = createAppContainer(DummyComponent, defaultApp);

    const component = renderer.create(<WrappedDummyComponent />);

    const firebaseAppProp = component.toJSON().props.firebaseApp;
    expect(firebaseAppProp).toEqual(defaultApp);
  });
});
