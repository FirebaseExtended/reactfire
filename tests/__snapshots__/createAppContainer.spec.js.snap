exports[`createAppContainer() should throw given an invalid Firebase App as the second argument 1`] = `"Second argument to createAppContainer() must be a Firebase app instance."`;

exports[`createAppContainer() should throw given an invalid React component as the first argument 1`] = `"First argument to createAppContainer() must be a React component."`;

exports[`createAppContainer() should throw given nothing as the first argument 1`] = `"First argument to createAppContainer() must be a React component."`;

exports[`createAppContainer() should throw given nothing as the second argument 1`] = `"Second argument to createAppContainer() must be a Firebase app instance."`;
